From 9bf11a4e5663a3c34756994b300e0bb1a49db38d Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Tue, 13 Feb 2024 22:14:16 +0530
Subject: [PATCH 26/34] SystemUI: VolumeDialog: Add app rows to volume dialog

Expand app rows instead of opening a settings panel.
Uses the per-app volume API.

Partially inspired by the patch from Android 11
"SystemUI: Add app volume row to volume dialog"
by cjybyjk <cjybyjk@zjnu.edu.cn>

Change-Id: I3728dfc4f39a06b96abfd571abb8858426e70f45
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../drawable/volume_row_seekbar_progress.xml  |  10 +-
 .../volume_row_seekbar_progress_icon.xml      |  21 ++
 .../systemui/volume/VolumeDialogImpl.java     | 319 ++++++++++++++----
 3 files changed, 277 insertions(+), 73 deletions(-)
 create mode 100644 packages/SystemUI/res/drawable/volume_row_seekbar_progress_icon.xml

diff --git a/packages/SystemUI/res/drawable/volume_row_seekbar_progress.xml b/packages/SystemUI/res/drawable/volume_row_seekbar_progress.xml
index fa06bd6ab..9977620c9 100644
--- a/packages/SystemUI/res/drawable/volume_row_seekbar_progress.xml
+++ b/packages/SystemUI/res/drawable/volume_row_seekbar_progress.xml
@@ -35,11 +35,7 @@
         android:right="@dimen/volume_slider_icon_inset">
         <rotate
             android:fromDegrees="-270"
-            android:toDegrees="-270">
-            <!-- A placeholder drawable is required here - it'll be replaced in code. -->
-            <com.android.systemui.util.AlphaTintDrawableWrapper
-                android:drawable="@drawable/ic_volume_media"
-                android:tint="?androidprv:attr/colorAccentPrimaryVariant" />
-        </rotate>
+            android:toDegrees="-270"
+            android:drawable="@drawable/volume_row_seekbar_progress_icon" />
     </item>
-</layer-list>
\ No newline at end of file
+</layer-list>
diff --git a/packages/SystemUI/res/drawable/volume_row_seekbar_progress_icon.xml b/packages/SystemUI/res/drawable/volume_row_seekbar_progress_icon.xml
new file mode 100644
index 000000000..74fce4785
--- /dev/null
+++ b/packages/SystemUI/res/drawable/volume_row_seekbar_progress_icon.xml
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (C) 2021 The Android Open Source Project
+  ~
+  ~ Licensed under the Apache License, Version 2.0 (the "License");
+  ~ you may not use this file except in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~      http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<com.android.systemui.util.AlphaTintDrawableWrapper
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:androidprv="http://schemas.android.com/apk/prv/res/android"
+    android:drawable="@drawable/ic_volume_media"
+    android:tint="?androidprv:attr/colorAccentPrimaryVariant" />
diff --git a/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java b/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java
index 57153719c..4d826ac17 100644
--- a/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java
@@ -54,6 +54,7 @@ import android.bluetooth.BluetoothProfile;
 import android.content.ContentResolver;
 import android.content.Context;
 import android.content.DialogInterface;
+import android.content.pm.ApplicationInfo;
 import android.content.pm.PackageManager;
 import android.content.res.ColorStateList;
 import android.content.res.Configuration;
@@ -61,6 +62,8 @@ import android.content.res.Resources;
 import android.content.res.TypedArray;
 import android.database.ContentObserver;
 import android.graphics.Color;
+import android.graphics.ColorMatrix;
+import android.graphics.ColorMatrixColorFilter;
 import android.graphics.Outline;
 import android.graphics.PixelFormat;
 import android.graphics.Rect;
@@ -199,6 +202,17 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
     /** Volume dialog slider animation. */
     private static final String TYPE_UPDATE = "update";
 
+    // Assign next available stream id for app volume
+    private static final int STREAM_APP = AudioSystem.getNumStreamTypes();
+
+    // Drawable color when the app is muted
+    private static final ColorMatrixColorFilter MONOCHROME_COLOR_FILTER;
+    static {
+        ColorMatrix matrix = new ColorMatrix();
+        matrix.setSaturation(0);
+        MONOCHROME_COLOR_FILTER = new ColorMatrixColorFilter(matrix);
+    }
+
     /**
      *  TODO(b/290612381): remove lingering animations or tolerate them
      *  When false, this will cause this class to not listen to animator events and not record jank
@@ -286,6 +300,7 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
     private View mAppVolumeView;
     private ImageButton mAppVolumeIcon;
     private final List<VolumeRow> mRows = new ArrayList<>();
+    private final List<VolumeRow> mAppRows = new ArrayList<>();
     private ConfigurableTexts mConfigurableTexts;
     private final SparseBooleanArray mDynamic = new SparseBooleanArray();
     private final KeyguardManager mKeyguard;
@@ -333,11 +348,20 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
     private FrameLayout mRoundedBorderBottom;
 
     // Volume panel expand state
-    private boolean mExpanded;
+    private enum ExpansionState {
+        COLLAPSED,
+        EXPANDED,
+        APPS_EXPANDED
+    }
+    private ExpansionState mExpansionState = ExpansionState.COLLAPSED;
+    private ExpansionState mPrevExpansionState = ExpansionState.COLLAPSED;
 
     // Number of animating rows
     private int mAnimatingRows = 0;
 
+    // Currently expanded app row
+    private String mActiveAppRowPackage;
+
     @VisibleForTesting
     final int mVolumeRingerIconDrawableId = R.drawable.ic_legacy_speaker_on;
     @VisibleForTesting
@@ -539,32 +563,37 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         float xExtraSize = 0;
         float yExtraSize = 0;
 
+        final boolean expanded = mExpansionState == ExpansionState.EXPANDED;
+        final boolean appsExpanded = mExpansionState == ExpansionState.APPS_EXPANDED;
+
         // The ringer and rows container has extra height at the top to fit the expanded ringer
         // drawer. This area should not be touchable unless the ringer drawer is open.
         // In landscape the ringer expands to the left and it has to be ensured that if there
         // are multiple rows they are touchable.
         // The invisible expandable rows reserve space if the panel is not expanded, this space
         // needs to be touchable.
+        // When app rows are expanded, stream rows are invisible and vice versa.
         if (view == mTopContainer) {
+            int extraRowsSize = 0;
+            if (!appsExpanded) {
+                extraRowsSize += getAppRowsExtraSize();
+            }
+            if (!expanded) {
+                extraRowsSize += getExpandableRowsExtraSize();
+            }
             if (!isLandscape()) {
                 if (!mIsRingerDrawerOpen) {
                     yExtraSize = getRingerDrawerOpenExtraSize();
                 }
-                if (!mExpanded) {
-                    xExtraSize = getExpandableRowsExtraSize();
-                }
+                xExtraSize = extraRowsSize;
             } else {
-                if (!mIsRingerDrawerOpen && !mExpanded) {
+                if (!mIsRingerDrawerOpen) {
                     xExtraSize =
-                            Math.max(getRingerDrawerOpenExtraSize(), getExpandableRowsExtraSize());
-                } else if (!mIsRingerDrawerOpen) {
-                    if (getRingerDrawerOpenExtraSize() > getVisibleRowsExtraSize()) {
-                        xExtraSize = getRingerDrawerOpenExtraSize() - getVisibleRowsExtraSize();
-                    }
-                } else if (!mExpanded) {
-                    if ((getVisibleRowsExtraSize() + getExpandableRowsExtraSize())
+                            Math.max(getRingerDrawerOpenExtraSize(), extraRowsSize);
+                } else {
+                    if ((getVisibleRowsExtraSize() + extraRowsSize)
                             > getRingerDrawerOpenExtraSize()) {
-                        xExtraSize = (getVisibleRowsExtraSize() + getExpandableRowsExtraSize())
+                        xExtraSize = (getVisibleRowsExtraSize() + extraRowsSize)
                                 - getRingerDrawerOpenExtraSize();
                     }
                 }
@@ -615,7 +644,7 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         mConfigurableTexts = new ConfigurableTexts(mContext);
         mHovering = false;
         mShowing = false;
-        mExpanded = false;
+        mExpansionState = mPrevExpansionState = ExpansionState.COLLAPSED;
         mWindow = mDialog.getWindow();
         mWindow.requestFeature(Window.FEATURE_NO_TITLE);
         mWindow.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
@@ -856,13 +885,14 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                         R.drawable.ic_volume_system_mute, false, false);
             }
         } else {
+            clearAppVolumes();
             addExistingRows();
         }
 
+        initAppVolumes();
         updateRowsH(getActiveRow());
         initRingerH();
         initSettingsH(lockTaskModeState);
-        initAppVolumeH();
         initODICaptionsH();
         mAccessibility.init();
     }
@@ -943,6 +973,38 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         }
     }
 
+    private void addAppRow(AppVolume av) {
+        ApplicationInfo appInfo;
+        PackageManager pm = mContext.getPackageManager();
+        String packageName = av.getPackageName();
+        if (D.BUG) Slog.d(TAG, "Adding row for app " + av.getPackageName());
+        try {
+            appInfo = pm.getApplicationInfo(packageName, 0);
+        } catch (PackageManager.NameNotFoundException e) {
+            Slog.e(TAG, "Unable to get app info for " + packageName);
+            return;
+        }
+        Drawable appIcon = pm.getApplicationIcon(appInfo);
+        String appLabel = pm.getApplicationLabel(appInfo).toString();
+        VolumeRow row = new VolumeRow();
+        row.isAppVolume = true;
+        row.appVolume = av;
+        row.packageName = packageName;
+        row.appLabel = appLabel;
+        row.appIcon = appIcon;
+        initRow(row, STREAM_APP, /*iconRes*/ 0, /*iconMuteRes*/ 0,
+                /*important*/ false, /*defaultStream*/ false);
+        Util.setText(row.header, row.appLabel);
+        row.slider.setContentDescription(row.appLabel);
+        row.slider.setEnabled(true);
+        row.slider.setMin(0);
+        row.slider.setMax(100);
+        mDialogRowsView.addView(row.view);
+        mRows.add(row);
+        mAppRows.add(row);
+        updateAppVolumeRow(row, true);
+    }
+
     private void addExistingRows() {
         int N = mRows.size();
         for (int i = 0; i < N; i++) {
@@ -956,6 +1018,9 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
 
     private VolumeRow getActiveRow() {
         for (VolumeRow row : mRows) {
+            if (row.isAppVolume && row.packageName.equals(mActiveAppRowPackage)) {
+                return row;
+            }
             if (row.stream == mActiveStream) {
                 return row;
             }
@@ -1032,10 +1097,19 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
 
         row.sliderProgressSolid = seekbarProgressDrawable.findDrawableByLayerId(
                 R.id.volume_seekbar_progress_solid);
-        final Drawable sliderProgressIcon = seekbarProgressDrawable.findDrawableByLayerId(
-                        R.id.volume_seekbar_progress_icon);
-        row.sliderProgressIcon = sliderProgressIcon != null ? (AlphaTintDrawableWrapper)
-                ((RotateDrawable) sliderProgressIcon).getDrawable() : null;
+
+        final RotateDrawable sliderProgressIcon = (RotateDrawable)
+                seekbarProgressDrawable.findDrawableByLayerId(R.id.volume_seekbar_progress_icon);
+
+        if (sliderProgressIcon == null) {
+            // do nothing
+        } else if (row.isAppVolume) {
+            sliderProgressIcon.setDrawable(row.appIcon);
+        } else {
+            row.sliderProgressIcon = (AlphaTintDrawableWrapper) mContext.getDrawable(
+                    R.drawable.volume_row_seekbar_progress_icon);
+            sliderProgressIcon.setDrawable(row.sliderProgressIcon);
+        }
 
         row.slider.setProgressDrawable(seekbarDrawable);
 
@@ -1044,7 +1118,7 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         row.setIcon(iconRes, mContext.getTheme());
 
         if (row.icon != null) {
-            if (row.stream != AudioSystem.STREAM_ACCESSIBILITY) {
+            if (row.stream != AudioSystem.STREAM_ACCESSIBILITY && !row.isAppVolume) {
                 row.icon.setOnClickListener(v -> {
                     Events.writeEvent(Events.EVENT_ICON_CLICK, row.stream, row.iconState);
                     mController.setActiveStream(row.stream);
@@ -1474,41 +1548,57 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         }
         if (mExpandRows != null) {
             mExpandRows.setOnClickListener(v -> {
-                mExpanded = !mExpanded;
+                // ensure we collapse app rows if expanded
+                if (mExpansionState == ExpansionState.APPS_EXPANDED) {
+                    mPrevExpansionState = mExpansionState;
+                    mExpansionState = ExpansionState.COLLAPSED;
+                    updateRowsH(mDefaultRow, false);
+                }
+                mPrevExpansionState = mExpansionState;
+                mExpansionState = mPrevExpansionState == ExpansionState.EXPANDED
+                        ? ExpansionState.COLLAPSED : ExpansionState.EXPANDED;
                 updateRowsH(mDefaultRow, true);
-                mExpandRows.setExpanded(mExpanded);
+                mExpandRows.setExpanded(mExpansionState == ExpansionState.EXPANDED);
             });
         }
     }
 
-    private boolean shouldShowAppVolume() {
-        ContentResolver cr = mContext.getContentResolver();
-        int showAppVolume = Settings.System.getInt(cr, Settings.System.SHOW_APP_VOLUME, 0);
-        boolean ret = showAppVolume == 1;
-        if (ret) {
-            ret = false;
-            AudioManager audioManager = mController.getAudioManager();
-            for (AppVolume av : audioManager.listAppVolumes()) {
-                if (av.isActive()) {
-                    ret = true;
-            break;
-                }
-            }
-        }
-        return ret;
+    private void clearAppVolumes() {
+        mActiveAppRowPackage = null;
+        mAppVolumeView.setVisibility(GONE);
+        mRows.removeAll(mAppRows);
+        mAppRows.forEach(row -> mDialogRowsView.removeView(row.view));
+        mAppRows.clear();
     }
 
-    public void initAppVolumeH() {
-        if (mAppVolumeView != null) {
-            mAppVolumeView.setVisibility(shouldShowAppVolume() ? VISIBLE : GONE);
+    private void initAppVolumes() {
+        clearAppVolumes();
+        boolean showAppVolume = Settings.System.getInt(mContext.getContentResolver(),
+                Settings.System.SHOW_APP_VOLUME, 0) == 1;
+        if (!showAppVolume || mAppVolumeView == null) {
+            return;
         }
-        if (mAppVolumeIcon != null) {
+        boolean appActive = false;
+        for (AppVolume av : mController.getAudioManager().listAppVolumes()) {
+            if (av.isActive()) {
+                addAppRow(av);
+                appActive = true;
+            }
+        }
+        if (appActive) {
+            mAppVolumeView.setVisibility(VISIBLE);
             mAppVolumeIcon.setOnClickListener(v -> {
-                Events.writeEvent(Events.EVENT_SETTINGS_CLICK);
-                dismissH(DISMISS_REASON_SETTINGS_CLICKED);
-                mMediaOutputDialogFactory.dismiss();
-                mVolumeNavigator.openVolumePanel(
-                        mVolumePanelNavigationInteractor.getAppVolumeRoute());
+                // ensure we collapse stream rows if expanded
+                if (mExpansionState == ExpansionState.EXPANDED) {
+                    mPrevExpansionState = mExpansionState;
+                    mExpansionState = ExpansionState.COLLAPSED;
+                    updateRowsH(mDefaultRow, false);
+                    mExpandRows.setExpanded(false);
+                }
+                mPrevExpansionState = mExpansionState;
+                mExpansionState = mPrevExpansionState == ExpansionState.APPS_EXPANDED
+                        ? ExpansionState.COLLAPSED : ExpansionState.APPS_EXPANDED;
+                updateRowsH(mDefaultRow, true);
             });
         }
     }
@@ -1780,6 +1870,11 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         mHandler.removeMessages(H.DISMISS);
         rescheduleTimeoutH();
 
+        if (mShowing && !mIsAnimatingDismiss) {
+            Log.i(TAG, "volume dialog is already showing");
+            return;
+        }
+
         if (mConfigChanged) {
             initDialog(lockTaskModeState); // resets mShowing to false
             mConfigurableTexts.update();
@@ -1790,8 +1885,9 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
             mDefaultRow = getActiveRow();
         }
 
+        initAppVolumes();
+        updateRowsH(mDefaultRow);
         initSettingsH(lockTaskModeState);
-        initAppVolumeH();
         mShowing = true;
         mIsAnimatingDismiss = false;
         mDialog.show();
@@ -1902,14 +1998,14 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                         mDialog.dismiss();
                     }
                     tryToRemoveCaptionsTooltip();
-                    mExpanded = false;
+                    mExpansionState = mPrevExpansionState = ExpansionState.COLLAPSED;
                     if (mExpandRows != null) {
-                        mExpandRows.setExpanded(mExpanded);
+                        mExpandRows.setExpanded(false);
                     }
+                    mActiveAppRowPackage = null;
                     mAnimatingRows = 0;
                     mDefaultRow = null;
                     mIsAnimatingDismiss = false;
-
                     hideRingerDrawer();
                 }, 50));
         if (!shouldSlideInVolumeTray()) {
@@ -1942,6 +2038,7 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                         || row.stream == STREAM_ALARM
                         || row.stream == STREAM_MUSIC);
     }
+
     private boolean shouldBeVisibleH(VolumeRow row, VolumeRow activeRow) {
         boolean isActive = row.stream == activeRow.stream;
 
@@ -1961,7 +2058,11 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                 return true;
             }
 
-            if (mExpanded && isExpandableRowH(row)) {
+            if (mExpansionState == ExpansionState.EXPANDED && isExpandableRowH(row)) {
+                return true;
+            }
+
+            if (mExpansionState == ExpansionState.APPS_EXPANDED && row.isAppVolume) {
                 return true;
             }
 
@@ -2005,7 +2106,13 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
             trimObsoleteH();
         }
 
-        boolean isOutmostIndexMax = isWindowGravityLeft() ? isRtl() : !isRtl();
+        final boolean expanded = mExpansionState == ExpansionState.EXPANDED
+                || mPrevExpansionState == ExpansionState.EXPANDED;
+        final boolean appsExpanded = mExpansionState == ExpansionState.APPS_EXPANDED
+                || mPrevExpansionState == ExpansionState.APPS_EXPANDED;
+        final boolean collapsed = mExpansionState == ExpansionState.COLLAPSED;
+
+        final boolean isOutmostIndexMax = isWindowGravityLeft() ? isRtl() : !isRtl();
 
         // Index of the last row that is actually visible.
         int outmostVisibleRowIndex = isOutmostIndexMax ? -1 : Short.MAX_VALUE;
@@ -2013,12 +2120,17 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         // apply changes to all rows
         for (final VolumeRow row : mRows) {
             final boolean isActive = row == activeRow;
-            final boolean isExpandableRow = isExpandableRowH(row);
+            final boolean isStreamRow = isExpandableRowH(row);
+            final boolean isAppRow = row.isAppVolume;
+            final boolean isExpandableRow = isStreamRow || isAppRow;
             final boolean shouldBeVisible = shouldBeVisibleH(row, activeRow);
+            // Hide stream rows when app rows expanded and vice versa
+            final boolean shouldBeInvisible =
+                    (isStreamRow && appsExpanded) || (isAppRow && expanded) || collapsed;
 
             if (!isExpandableRow) {
                 Util.setVisOrGone(row.view, shouldBeVisible);
-            } else if (!mExpanded) {
+            } else if (shouldBeInvisible) {
                 row.view.setVisibility(View.INVISIBLE);
             }
 
@@ -2089,7 +2201,11 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
             int rowWidth = mDialogWidth + mRingerRowsPadding;
 
             for (final VolumeRow row : mRows) {
-                final boolean isExpandableRow = isExpandableRowH(row);
+                final boolean isStreamRow = isExpandableRowH(row);
+                final boolean isAppRow = row.isAppVolume;
+                final boolean isExpandableRow = isStreamRow || isAppRow;
+                final boolean isExpanded = mExpansionState == ExpansionState.EXPANDED
+                        || mExpansionState == ExpansionState.APPS_EXPANDED;
                 final boolean shouldBeVisible = shouldBeVisibleH(row, activeRow);
 
                 if (shouldBeVisible) {
@@ -2108,14 +2224,19 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
 
                 // Expandable rows need to animate behind the last visible row, an additional
                 // always visible stream animates next to the default stream
-                float translationX = (isWindowGravityLeft() ? -1 : 1) * hiddenRowsCount * rowWidth;
+                final int translationMultiplier = isWindowGravityLeft() ? -1 : 1;
+                float fromTranslationX = translationMultiplier * visibleRowsCount * rowWidth;
+                float toTranslationX = translationMultiplier * hiddenRowsCount * rowWidth;
+                if (isAppRow) {
+                    // Move app rows to take over invisible expandable rows
+                    fromTranslationX += translationMultiplier * getExpandableRowsExtraSize();
+                }
 
                 if (animate) {
                     // If the translation is not set and the panel is expanding start the
                     // animation behind the main row
-                    if (isExpandableRow && mExpanded && row.view.getTranslationX() == 0f) {
-                        row.view.setTranslationX(
-                                (isWindowGravityLeft() ? -1 : 1) * visibleRowsCount * rowWidth);
+                    if (isExpandableRow && isExpanded && row.view.getTranslationX() == 0f) {
+                        row.view.setTranslationX(fromTranslationX);
                     }
 
                     // The elevation should decrease from the outmost row to the inner rows, so that
@@ -2128,10 +2249,10 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                     // if there is still a row animating
                     mAnimatingRows++;
                     row.view.animate()
-                            .translationX(translationX)
-                            .setDuration(mExpanded ? mDialogShowAnimationDurationMs
+                            .translationX(toTranslationX)
+                            .setDuration(isExpanded ? mDialogShowAnimationDurationMs
                                                    : mDialogHideAnimationDurationMs)
-                            .setInterpolator(mExpanded
+                            .setInterpolator(isExpanded
                                             ? new SystemUIInterpolators.LogDecelerateInterpolator()
                                             : new SystemUIInterpolators.LogAccelerateInterpolator())
                             .setListener(new AnimatorListenerAdapter() {
@@ -2157,11 +2278,12 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                                         // already, but setting the active row cancels ongoing
                                         // animations.
                                         mController.setActiveStream(activeRow.stream);
+                                        mActiveAppRowPackage = null;
                                     }
                                 }
                             });
                 } else {
-                    row.view.setTranslationX(translationX);
+                    row.view.setTranslationX(toTranslationX);
                     row.view.setVisibility(shouldBeVisible ? View.VISIBLE : View.INVISIBLE);
                 }
             }
@@ -2314,6 +2436,7 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         if (mActiveStream != state.activeStream) {
             mPrevActiveStream = mActiveStream;
             mActiveStream = state.activeStream;
+            mActiveAppRowPackage = null;
             VolumeRow activeRow = getActiveRow();
             updateRowsH(activeRow);
             if (mShowing) rescheduleTimeoutH();
@@ -2332,6 +2455,10 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
 
     private void updateVolumeRowH(VolumeRow row) {
         if (D.BUG) Log.i(TAG, "updateVolumeRowH s=" + row.stream);
+        if (row.isAppVolume) {
+            updateAppVolumeRow(row);
+            return;
+        }
         if (mState == null) return;
         final StreamState ss = mState.states.get(row.stream);
         if (ss == null) return;
@@ -2471,6 +2598,32 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         if (row.number != null) row.number.setText(Integer.toString(vlevel));
     }
 
+    private void updateAppVolumeRow(VolumeRow row) {
+        updateAppVolumeRow(row, false);
+    }
+
+    private void updateAppVolumeRow(VolumeRow row, boolean setProgress) {
+        if (D.BUG) Slog.d(TAG, "updateAppVolumeRow: packageName=" + row.packageName);
+
+        // update slider
+        updateVolumeRowTintH(row, row.packageName.equals(mActiveAppRowPackage));
+        if (setProgress) {
+            final int progress = Math.round(row.appVolume.getVolume() * 100f);
+            row.slider.setProgress(progress);
+        }
+
+        // update icon
+        if (row.slider.getProgress() == 0 || row.appVolume.isMuted()) {
+            row.appIcon.setColorFilter(MONOCHROME_COLOR_FILTER);
+        } else {
+            row.appIcon.clearColorFilter();
+        }
+    }
+
+    private void updateAppVolumeRows() {
+        mAppRows.forEach(this::updateAppVolumeRow);
+    }
+
     private boolean isStreamMuted(final StreamState streamState) {
         return (mAutomute && streamState.level == 0) || streamState.muted;
     }
@@ -2723,7 +2876,6 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
      * Expandable rows are invisible while the panel is not expanded.
      */
     private int getExpandableRowsExtraSize() {
-        VolumeRow activeRow = getActiveRow();
         int expandableRows = 0;
         for (final VolumeRow row : mRows) {
             if (isExpandableRowH(row)) {
@@ -2733,6 +2885,16 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         return expandableRows * (mDialogWidth + mRingerRowsPadding);
     }
 
+    private int getAppRowsExtraSize() {
+        int appRows = 0;
+        for (final VolumeRow row : mRows) {
+            if (row.isAppVolume) {
+                appRows++;
+            }
+        }
+        return appRows * (mDialogWidth + mRingerRowsPadding);
+    }
+
     private void updateBackgroundForDrawerClosedAmount() {
         if (mRingerAndDrawerContainerBackground == null) {
             return;
@@ -3018,7 +3180,6 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
 
         @Override
         public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
-            if (mRow.ss == null) return;
             if (getActiveRow().equals(mRow) && mRow.slider.getVisibility() == VISIBLE) {
                 if (fromUser || mRow.animTargetProgress == progress) {
                     // Deliver user-generated slider haptics immediately, or when the animation
@@ -3029,6 +3190,15 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
             if (D.BUG) Log.d(TAG, AudioSystem.streamToString(mRow.stream)
                     + " onProgressChanged " + progress + " fromUser=" + fromUser);
             if (!fromUser) return;
+            if (mRow.isAppVolume) {
+                mActiveAppRowPackage = mRow.packageName;
+                final float vol = progress * 0.01f;
+                if (D.BUG) Log.d(TAG, "set app " + mRow.packageName + " volume to " + vol);
+                mController.getAudioManager().setAppVolume(mRow.packageName, vol);
+                updateAppVolumeRows();
+                return;
+            }
+            if (mRow.ss == null) return;
             if (mRow.ss.levelMin > 0) {
                 final int minProgress = getProgressFromVolume(mRow.ss, seekBar, mRow.ss.levelMin);
                 if (progress < minProgress) {
@@ -3040,6 +3210,7 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
             if (mRow.ss.level != userLevel || mRow.ss.muted && userLevel > 0) {
                 mRow.userAttempt = SystemClock.uptimeMillis();
                 if (mRow.requestedLevel != userLevel) {
+                    mActiveAppRowPackage = null;
                     mController.setActiveStream(mRow.stream);
                     mController.setStreamVolume(mRow.stream, userLevel);
                     mRow.requestedLevel = userLevel;
@@ -3056,8 +3227,14 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
             if (mRow.mHapticPlugin != null) {
                 mRow.mHapticPlugin.onStartTrackingTouch();
             }
-            mController.setActiveStream(mRow.stream);
             mRow.tracking = true;
+            if (mRow.isAppVolume) {
+                mActiveAppRowPackage = mRow.packageName;
+                updateAppVolumeRows();
+                return;
+            }
+            mController.setActiveStream(mRow.stream);
+            mActiveAppRowPackage = null;
         }
 
         @Override
@@ -3068,6 +3245,10 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                 mRow.mHapticPlugin.onStopTrackingTouch();
             }
             mRow.tracking = false;
+            if (mRow.isAppVolume) {
+                updateAppVolumeRows();
+                return;
+            }
             mRow.userAttempt = SystemClock.uptimeMillis();
             final int userLevel = getVolumeFromProgress(mRow.ss, seekBar, seekBar.getProgress());
             Events.writeEvent(Events.EVENT_TOUCH_LEVEL_DONE, mRow.stream, userLevel);
@@ -3145,6 +3326,12 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
         private int animTargetProgress;
         private int lastAudibleLevel = 1;
         private HapticSliderPlugin mHapticPlugin;
+        // per-app volume
+        private boolean isAppVolume = false;
+        private AppVolume appVolume;
+        private String packageName;
+        private String appLabel;
+        private Drawable appIcon;
 
         void setIcon(int iconRes, Resources.Theme theme) {
             if (icon != null) {
-- 
2.52.0

